OpenRA.Mods.Common/Traits/World/VariedCostManager.cs

```diff
@@ -0,0 +1,78 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Attach this to the world actor.")]
+	public class VariedCostManagerInfo : TraitInfo
+	{
+		[Desc("Interval between new pricings measured in ticks.")]
+		public readonly int AdjustmentDelay = 1000;
+
+		[Desc("Range of percentage modifiers to apply.")]
+		public readonly int2 Multiplier = new int2(75, 100);
+
+		public override object Create(ActorInitializer init) { return new VariedCostManager(this, init.Self); }
+	}
+
+	public class VariedCostManager : ITick
+	{
+		public readonly Dictionary<string, int> CachedCostPercentage = new Dictionary<string, int>();
+
+		readonly VariedCostManagerInfo info;
+		readonly Dictionary<string, string> actorsByGroup = new Dictionary<string, string>();
```

Comment:
The name `actorsByGroup` means that you have a dictionary with actor values that are keyed by their group. This appears to be backwards to how it is actually used.

It's not clear to me why we need three different dictionaries to track this at all? All we should need is a single dictionary of group &rarr; modifier, where the group is evaluated as `variedCostMultiplierInfo.Group ?? actorInfo.Key`. The Tick loop can then loop over the keys and update the corresponding value.

This dictionary should also remain private, with a `public int GetCostModifier(string group)` method exposed for `VariedCostMultiplierInfo` to access.

Reply:
`actorsByGroup` is for combat tanks which are one cost group.

Reply:
How can `VariedCostMultiplierInfo` be used to access non-static values?

Reply:
I don't understand the question. The same way you are accessing non-static values now?

Reply:
I mean the multiplier changes at runtime. TraitInfo is used for configuration values that are read from file.